libs/injection/bbhnet/injection/injection.py

```diff
@@ -0,0 +1,261 @@
+):
+
+    """ Start simulation """
+
+    strain = TimeSeries.read(frame_files[0], channels[0])
+    logging.info(f"Read strain from {frame_files[0]}")
+
+    frame_start, frame_stop = strain.span
+    frame_duration = frame_stop - frame_start
+
+    sample_rate = int(strain.sample_rate.value)
+    fftlength = int(max(2, np.ceil(2048 / sample_rate)))
+
+    # set the non-overlapping times of the signals in the frames randomly
+    # leaves buffer at either end of the series so edge effects aren't an issue
+    signal_times = sorted(
+        np.random.choice(
+            np.arange(
+                waveform_duration,
+                frame_duration - waveform_duration,
+                waveform_duration,
+            ),
+            size=n_samples,
+            replace=False,
+        )
+    )
+
+    # log and print out some simulation parameters
+    logging.info("Simulation parameters")
+    logging.info("Number of samples     : {}".format(n_samples))
+    logging.info("Sample rate [Hz]      : {}".format(sample_rate))
+    logging.info("High pass filter [Hz] : {}".format(fmin))
+    logging.info("Prior file            : {}".format(prior_file))
+
+    # define a Bilby waveform generator
+    waveform_generator = bilby.gw.WaveformGenerator(
+        duration=waveform_duration,
+        sampling_frequency=sample_rate,
+        frequency_domain_source_model=lal_binary_black_hole,
+        parameter_conversion=convert_to_lal_binary_black_hole_parameters,
+        waveform_arguments={
+            "waveform_approximant": "IMRPhenomPv2",
+            "reference_frequency": 50,
+            "minimum_frequency": 20,
+        },
+    )
+
+    # sample GW parameters from prior distribution
+    priors = bilby.gw.prior.BBHPriorDict(prior_file)
+    sample_params = priors.sample(n_samples)
+
+    # Come back to this and see if can add frame_start
+    sample_params["geocent_time"] = signal_times
+
+    signal_strains = []
+    signals_list = []
+    snr_list = []
+    for frame, channel, ifo in zip(frame_files, channels, ifos):
+        # Get the strain from each frame file
+        strain = TimeSeries.read(frame, channel)
+
+        # calculate the PSD
+        strain_psd = strain.psd(fftlength)
+
+        # generate GW waveforms
+        signals, snr = generate_gw(
+            sample_params,
+            ifo,
+            waveform_generator=waveform_generator,
+            get_snr=True,
+            noise_psd=strain_psd,
+        )
+
+        signal_strains.append(strain)
+        signals_list.append(signals)
+        snr_list.append(snr)
+
+    old_snr = np.sqrt(np.sum(np.square(snr_list)))
+    new_snr = np.random.uniform(snr_range[0], snr_range[1], len(snr_list[0]))
+
+    signals_list = [
+        signals * (new_snr / old_snr)[:, None] for signals in signals_list
+    ]
+    sample_params["luminosity_distance"] = (
+        sample_params["luminosity_distance"] * old_snr / new_snr
+    )
+    snr_list = [snr * new_snr / old_snr for snr in snr_list]
```

Comment:
Can you explain what this block of code is doing? Do we need to do it before we write the injections to files? I can think of a lot of reasons that it would be nice to be able write injections as soon as we've created them, and not have to hold them all into memory at once.
- Allows us to do a lot more injections at once
- Allows us to not lose any data if anything goes wrong
- Saves us the extra loop below

Reply:
Decided on desiring functionality for injecting sets of frames, so left as-is